src/detect-yara.c

```diff
@@ -0,0 +1,435 @@
+            SCLogError(SC_WARN_YARA_ERROR, "too many yara string matches");
+            break;
+        default:
+            SCLogError(SC_WARN_YARA_ERROR, "unknown yara error: %d", error);
+            break;
+    }
+}
+
+/**
+ * \brief count yara matches
+ *
+ * \param message int error type
+ * \param message_data void not used
+ * \param user_data void match counter
+ *
+ * \retval CALLBACK_CONTINUE
+ */
+static int YaraCallback(int message, void* message_data, void* user_data)
+{
+    if (message == CALLBACK_MSG_RULE_MATCHING)
+        (*(int *)user_data)++;
+
+    return CALLBACK_CONTINUE;
+}
+
+/**
+ * \brief match the specified yara rule
+ *
+ * \param t thread local vars
+ * \param det_ctx pattern matcher thread local data
+ * \param f *LOCKED* flow
+ * \param flags direction flags
+ * \param file file being inspected
+ * \param s signature being inspected
+ * \param m sigmatch that we will cast into DetectYaraData
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectYaraMatch(ThreadVars *t, DetectEngineThreadCtx *det_ctx, Flow *f,
+        uint8_t flags, File *file, const Signature *s, const SigMatchCtx *m)
+{
+    SCEnter();
+    int result = 0;
+    int matches = 0;
+    DetectYaraData *yara = (DetectYaraData *)m;
+
+    const uint8_t *data = NULL;
+    uint32_t data_len = 0;
+    uint64_t offset = 0;
+    uint64_t file_size = FileTrackedSize(file);
+    StreamingBufferGetData(file->sb, &data, &data_len, &offset);
+
+    /* Perform the YARA scan. Require at least one YARA match */
+    result = yr_rules_scan_mem(yara->rules, data, file_size, yara_fast_scan,
+                               YaraCallback, &matches, yara_timeout);
+
+    if (result == ERROR_SUCCESS) {
+        if (matches >= 1)
+            SCReturnInt(yara->negated ? 0 : 1);
+        else
+            SCReturnInt(yara->negated ? 1 : 0);
+    } else {
+        YaraError(result);
+        SCReturnInt(0);
+    }
+
+    SCReturnInt(0);
+}
+
+/**
+ * \brief This function is used to parse yara options passed via yara: keyword
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param str Pointer to the user provided yara options
+ * \param negate flag to indicate if the rule is negated
+ *
+ * \retval yara pointer to DetectYaraData on success
+ * \retval NULL on failure
+ */
+static DetectYaraData *DetectYaraParse(const DetectEngineCtx *de_ctx,
+        const char *str, int negate)
+{
+    SCEnter();
+    DetectYaraData *yara = NULL;
+
+    yara = SCMalloc(sizeof(DetectYaraData));
+    if (unlikely(yara == NULL))
+        goto error;
+
+    memset(yara, 0x00, sizeof(DetectYaraData));
+
+    if (negate)
+        yara->negated = 1;
+
+    /* get full filename */
+    yara->filename = DetectLoadCompleteSigPath(de_ctx, str);
+    if (yara->filename == NULL)
+        goto error;
+
+    return yara;
+
+error:
+    if (yara != NULL)
+        DetectYaraFree(yara);
+    if (yara->filename != NULL)
+        SCFree(yara->filename);
+    return NULL;
```

Comment:
yara has already been freed here. Also, it looks like DetectYaraFree already free's the filename.